### ISSUE TITLE
Relative extraPaths values allow importing files once but not transitively

### ISSUE THREAD

**Issue by ipkiss42** [microsoft/pyright#1515]
**Describe the bug**
Relative `extraPaths` values allow importing files once but not transitively

**To Reproduce**
Create the following file structure:

```
.
├── root1
│   ├── pyrightconfig.json
│   └── src
│       └── foo
│           ├── foo1.py
│           └── __init__.py
└── root2
    └── bar
        ├── bar1.py
        ├── bar2.py
        └── __init__.py
```

`root1/pyrightconfig.json`:

```json
{
    "executionEnvironments": [
        {
            "root": "src",
            "extraPaths": [
                "../root2"
            ]
        }
    ]
}
```

`root1/src/foo/foo1.py`:

```python
from bar import bar1
```

`root2/bar/bar1.py`:

```python
from bar import bar2
```

Other files are empty.

**Expected behavior**
No error.

**Screenshots or Code**

Pyright output (from `root1`):

```bash
$ pyright --verbose
Loading configuration file at /tmp/pyright1/root1/pyrightconfig.json
Assuming Python version 3.8
Assuming Python platform Linux
No include entries specified; assuming /tmp/pyright1/root1
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding .git
Search paths found for configured python interpreter:
  /tmp/python-3.8/lib/python3.8
  /tmp/python-3.8/lib/python3.8/lib-dynload
  /tmp/python-3.8/lib/python3.8/site-packages
When attempting to get search paths from python interpreter:
  Executing interpreter: 'python3'
  Skipping '/tmp/python-3.8/lib/python38.zip' because it is not a valid directory
  Received 3 paths from interpreter
    /tmp/python-3.8/lib/python3.8
    /tmp/python-3.8/lib/python3.8/lib-dynload
    /tmp/python-3.8/lib/python3.8/site-packages
stubPath /tmp/pyright1/root1/typings is not a valid directory.
Searching for source files
Found 2 source files
Could not import 'bar' in file '/tmp/pyright1/root2/bar/bar1.py'
  Looking for typeshed stdlib path
  Attempting to resolve using root path '/tmp/node_modules/pyright/dist/typeshed-fallback/stdlib'
  Typeshed path not found
  Looking in stubPath '/tmp/pyright1/root1/typings'
  Attempting to resolve using root path '/tmp/pyright1/root1/typings'
  Looking in root directory of execution environment '/tmp/pyright1/root1'
  Attempting to resolve using root path '/tmp/pyright1/root1'
  Looking in python search path '/tmp/python-3.8/lib/python3.8'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8'
  Looking in python search path '/tmp/python-3.8/lib/python3.8/lib-dynload'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8/lib-dynload'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8/lib-dynload'
  Looking in python search path '/tmp/python-3.8/lib/python3.8/site-packages'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8/site-packages'
  Attempting to resolve using root path '/tmp/python-3.8/lib/python3.8/site-packages'
  Looking for typeshed path
  Looking for typeshed stubs path
  Typeshed path not found
0 errors, 0 warnings, 0 infos
Completed in 0.523sec
```

**Additional context**

I also noticed the following:
* The import is resolved properly if the extra path is absolute instead of relative.
* If the `foo` module is moved up to `root1` (and the `root` is updated to `.` in `pyrightconfig.json`), the problem also disappears.

**Comment by erictraut** [comment]
The `extraPaths` is meant to specify subdirectories within your workspace — i.e. directories under the root. You're specifying paths outside of your workspace. Pyright should probably detect and log an error in this case and ignore the incorrect path. I'm thinking that we should also maybe disallow absolute paths. They should never be used for `extraPaths`. I'm noticing that a bunch of users are trying to use absolute paths here, and that was never meant to be supported.

@jakebailey, any thoughts here?

**Comment by jakebailey** [comment]
Absolute extraPaths are something we intended to support (and did in MPLS). This sort of thing is frequently done in cases where the code is going to be run in a different environment, the most common I've seen are for blender plugins (but there are others). It shouldn't be any different than using say an editable installs that points outside the workspace.

I'm not so sure about this case where it's relative and leaving the workspace, but I also don't see a reason to ban it. Could be an issue with my modifications fixing relative paths (can look Monday when I'm available; busy weekend....)

**Comment by jakebailey** [comment]
I think this is just like #1522; IIRC extraPaths, if relative, is relative to the workspace root / execution environment root. Can you try `../../root2` with this case to verify?

**Comment by jakebailey** [comment]
Whether or not that behavior is "correct", I'm unsure; I think I have to re-look at the settings code to figure this out. A long-term item I'd like to see happen is for the configuration code to be refactored so that all of this calculation is done early (and then is just handed to the analysis code). Hard to change this, as there's always someone depending on it, and this relative behavior makes the most sense for VS Code multi-root.

**Comment by erictraut** [comment]
The problem with absolute paths is that the pyrightconfig.json file is intended to be checked in to a repo and usable by anyone who checks out the project. If absolute paths are allowed, it breaks this. The same is true for relative paths outside of the project. I'm still not convinced we should support this.

**Comment by erictraut** [comment]
This turns out to be the same issue as #1523

**Comment by ipkiss42** [comment]
> The problem with absolute paths is that the pyrightconfig.json file is intended to be checked in to a repo and usable by anyone who checks out the project. If absolute paths are allowed, it breaks this. The same is true for relative paths outside of the project. I'm still not convinced we should support this.

Whether or not you support it is of course up to you, but from my perspective this assumed "intent" is an arbitrary limitation for your users.

For example, in my case the pyrightconfig.json is generated so I don't need to check it in. As a result, having the ability to use (relative or absolute) paths outside of the workspace is convenient, and I didn't even consider it might not work. You might argue that generating the file is an abuse of the system, but it's not clear to me why Pyright should care about that.

Out of curiosity, how would you support a multi-root workspace (like in VS Code) with one or several pyrightconfig.json file(s)?
* Having one pyrightconfig.json per root seems to imply that some paths will point outside of the projects.
* Having one pyrightconfig.json for everything requires that it lives in a directory which is a common parent of all the root folders. In the worst case, this common parent might be `/` (and it is not a convenient place to check it in a repo).
* The last option I can think of is to play with symlinks to trick Pyright into seeing paths within the project, when they are actually outside. I haven't checked whether this would work, but this would be a hack, to workaround the above limitation.

**Comment by erictraut** [comment]
Pyright/pylance will instantiate the language server once for each workspace in a multi-root workspace. It's as if you had multiple independent copies of the type analyzer running. That means each one has independent settings and configuration files, import paths, etc. You would need one pyrightconfig.json per root.

Yeah, I see your point about how relative and absolute paths would be useful here. They are supported today, and we don't have any need to remove that support. It's just not what I had in mind when I designed the "execution environment" capability in the pyrightconfig.json file.

**Comment by erictraut** [comment]
See #1508 for the resolution we've decided upon. In summary, transitive imports will not be supported with execution environments. If you need that capability, you can create separate pyrightconfig.json files, one for each root.
